wal/src/message_queue_impl/region.rs

```diff
@@ -0,0 +1,761 @@
+    async fn mark_delete_up_to(
+        &self,
+        table_id: TableId,
+        sequence_num: SequenceNumber,
+    ) -> Result<()> {
+        debug!(
+            "Mark deleted entries up to sequence_num:{}, region_id:{}",
+            sequence_num, self.region_id
+        );
+
+        self.region_meta
+            .mark_table_deleted(table_id, sequence_num)
+            .await
+            .context(MarkDeleted {
+                region_id: self.region_id,
+                table_id,
+            })
+    }
+
+    /// Get meta data by table id.
+    async fn get_table_meta(&self, table_id: TableId) -> Result<TableMetaData> {
+        self.region_meta
+            .get_table_meta_data(table_id)
+            .await
+            .context(GetTableMeta {
+                region_id: self.region_id,
+                table_id,
+            })
+    }
+
+    /// Get meta data snapshot of whole region.
+    ///
+    /// NOTICE: should freeze whole region before calling.
+    async fn make_meta_snapshot(&self) -> RegionMetaSnapshot {
+        self.region_meta.make_snapshot().await
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug)]
+pub struct MessageQueueLogIterator<C: ConsumeIterator> {
+    /// Id of region
+    region_id: RegionId,
+
+    /// Polling's end point
+    ///
+    /// While fetching in slave node, it will be set to `None`, and
+    /// reading will not stop.
+    /// Otherwise, it will be set to high watermark.
+    terminate_offset: Option<Offset>,
+
+    /// Terminated flag
+    is_terminated: bool,
+    /// Consume Iterator of message queue
+    iter: C,
+    /// Used to encode/decode the logs
+    log_encoding: CommonLogEncoding,
+    /// See the same problem in https://github.com/CeresDB/ceresdb/issues/120
+    previous_value: Vec<u8>,
+    // TODO: timeout
+}
+
+#[allow(unused)]
+impl<C: ConsumeIterator> MessageQueueLogIterator<C> {
+    fn new(
+        region_id: RegionId,
+        terminate_offset: Option<Offset>,
+        iter: C,
+        log_encoding: CommonLogEncoding,
+    ) -> Self {
+        Self {
+            region_id,
+            terminate_offset,
+            iter,
+            is_terminated: false,
+            log_encoding,
+            previous_value: Vec::new(),
+        }
+    }
+}
+
+#[allow(unused)]
+impl<C: ConsumeIterator> MessageQueueLogIterator<C> {
+    pub async fn next_log_entry(&mut self) -> Result<Option<LogEntry<&'_ [u8]>>> {
+        if self.is_terminated && self.terminate_offset.is_some() {
+            info!(
+                "Finished to poll log from message queue, region id:{}, terminate offset:{:?}",
+                self.region_id, self.terminate_offset
+            );
+            return Ok(None);
+        }
+
+        let (message_and_offset, high_watermark) = self.iter.next_message().await.unwrap();
+
+        // Check has race happened.
+        if let Some(terminate_offset) = &self.terminate_offset {
+            ensure!(*terminate_offset == high_watermark, ReadAllLogsNoCause {
```

Comment:
I can't see where this condition is ensured to be true.

Reply:
Yes, it should not be the logic here, because it is a basic component used for `scanning region during recovering`, `scanning the range of a table`, `fetching for slave node`. 

The only restriction for the `terminate_offset ` should be less than or equal to high watermark.

I plan to define some high-level iterator wrapping it as the inner later, for processing the high-level logic. But the pr is a bit larger, so let us keep it now